llama_stack/providers/inline/vector_io/sqlite_vec/sqlite_vec.py

```
@@ -80,21 +82,35 @@ async def add_chunks(self, chunks: List[Chunk], embeddings: NDArray):
         try:
             # Start transaction
             cur.execute("BEGIN TRANSACTION")
-            for chunk, emb in zip(chunks, embeddings, strict=False):
-                # Serialize and insert the chunk metadata.
-                chunk_json = chunk.model_dump_json()
-                cur.execute(f"INSERT INTO {self.metadata_table} (chunk) VALUES (?)", (chunk_json,))
-                row_id = cur.lastrowid
-                # Ensure the embedding is a list of floats.
-                emb_list = emb.tolist() if isinstance(emb, np.ndarray) else list(emb)
-                emb_blob = serialize_vector(emb_list)
-                cur.execute(f"INSERT INTO {self.vector_table} (rowid, embedding) VALUES (?, ?)", (row_id, emb_blob))
-            # Commit transaction if all inserts succeed
+            for i in range(0, len(chunks), batch_size):
+                batch_chunks = chunks[i : i + batch_size]
+                batch_embeddings = embeddings[i : i + batch_size]
+                # Prepare metadata inserts
+                metadata_data = [
+                    (generate_chunk_id(chunk.metadata["document_id"], chunk.content), chunk.model_dump_json())
+                    for chunk in batch_chunks
+                ]
+                # Insert metadata (ON CONFLICT to avoid duplicates)
+                cur.executemany(
+                    f"""
+                    INSERT INTO {self.metadata_table} (id, chunk)
+                    VALUES (?, ?)
+                    ON CONFLICT(id) DO UPDATE SET chunk = excluded.chunk;
```

We shouldn't expect conflicts right? Better to throw an error?

---

If a user were to submit the same chunk twice, for whatever reason, we'd just update the chunk. I think that's a better UX than raising an error, no?

---

e.g., if a user accidentally duplicates chunks we'll only write one value.

---

I can throw the error if you want though, let me know what you think @ehhuang. Thanks for the review!

---

Yeah, agree that it would be easier to update than raise. I can see us rerunning the same script multiple times ( we currently have a uuid in the name of the vector to avoid this